From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
Date: Sun, 23 Jun 2019 17:01:48 +0200
Subject: [PATCH] {PRAGMA} arch/x86/kernel/head32: ignore memset crossing
 member bounds
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

gcc on x86-32 reports:

      CC      arch/x86/kernel/head32.o
    In file included from arch/x86/include/asm/page_32.h:35,
                     from arch/x86/include/asm/page.h:14,
                     from arch/x86/include/asm/thread_info.h:12,
                     from include/linux/thread_info.h:38,
                     from arch/x86/include/asm/preempt.h:7,
                     from include/linux/preempt.h:78,
                     from include/linux/spinlock.h:51,
                     from include/linux/mmzone.h:8,
                     from include/linux/gfp.h:6,
                     from include/linux/mm.h:10,
                     from arch/x86/kernel/head32.c:11:
    In function ‘memset’,
        inlined from ‘sanitize_boot_params’ at arch/x86/include/asm/bootparam_utils.h:40:3,
        inlined from ‘i386_start_kernel’ at arch/x86/kernel/head32.c:39:2:
    include/linux/string.h:356:9: error: ‘__builtin_memset’ offset [197,
    448] from the object at ‘boot_params’ is out of the bounds of
    referenced subobject ‘ext_ramdisk_image’ with type ‘unsigned int’ at
    offset 192 [-Werror=array-bounds]
      356 |  return __builtin_memset(p, c, size);
          |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
    In function ‘memset’,
        inlined from ‘sanitize_boot_params’ at arch/x86/include/asm/bootparam_utils.h:43:3,
        inlined from ‘i386_start_kernel’ at arch/x86/kernel/head32.c:39:2:
    include/linux/string.h:356:9: error: ‘__builtin_memset’ offset [493,
    497] from the object at ‘boot_params’ is out of the bounds of
    referenced subobject ‘kbd_status’ with type ‘unsigned char’ at
    offset 491 [-Werror=array-bounds]
      356 |  return __builtin_memset(p, c, size);
          |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
---
 arch/x86/include/asm/bootparam_utils.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/x86/include/asm/bootparam_utils.h b/arch/x86/include/asm/bootparam_utils.h
index 101eb944f13c..9de0b0e7fb22 100644
--- a/arch/x86/include/asm/bootparam_utils.h
+++ b/arch/x86/include/asm/bootparam_utils.h
@@ -35,6 +35,9 @@ static void sanitize_boot_params(struct boot_params *boot_params)
 	 * problems again.
 	 */
 	if (boot_params->sentinel) {
+/* Some of theses memset() calls cross the boundary of the members */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Warray-bounds"
 		/* fields in boot_params are left uninitialized, clear them */
 		boot_params->acpi_rsdp_addr = 0;
 		memset(&boot_params->ext_ramdisk_image, 0,
@@ -50,6 +53,7 @@ static void sanitize_boot_params(struct boot_params *boot_params)
 		       (char *)&boot_params->eddbuf[0] -
 			(char *)&boot_params->_pad8[0]);
 		memset(&boot_params->_pad9[0], 0, sizeof(boot_params->_pad9));
+#pragma GCC diagnostic pop
 	}
 }
 
-- 
